### ISSUE TITLE
Batch Normalization with virtual_batch_size not equal to None not implemented correctly for inference time

### ISSUE THREAD

**Issue by ccurro** [tensorflow/tensorflow#23050]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.10.1-0-g4dcfddc5d1 1.10.1
- **Python version**: Python 3.6.2
- **Exact command to reproduce**: python main.py
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Mobile device**: N/A
- **Bazel version**: N/A

### Describe the problem
The batch normalization implementation respects the virtual_batch_size parameter in both train and inference modes. As such you are unable to do inference with batch sizes that are not multiples of the virtual_batch_size. Algorithm 1 in the [ghost batch norm paper](https://arxiv.org/pdf/1705.08741.pdf) makes it clear that virtual_batch_size should only be respected in train mode.

Relevant tf source code: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/normalization.py

### Source code / logs

```python
import tensorflow as tf
import numpy as np

virtual_batch_size = 32

training = tf.placeholder(tf.bool, shape=[])
x = tf.placeholder(tf.float32, shape=[None, 28, 28, 1])
x_norm = tf.layers.batch_normalization(x, training=training, virtual_batch_size=virtual_batch_size)
update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)

sess = tf.Session()

sess.run(tf.global_variables_initializer())


for _ in range(10):
    sess.run([update_ops, x_norm], feed_dict={x : np.random.random(size=[4*virtual_batch_size, 28, 28, 1]), training : True})

sess.run(x_norm, feed_dict={x : np.random.random(size=[virtual_batch_size, 28, 28, 1]), training : False})

# fails after this line
sess.run(x_norm, feed_dict={x : np.random.random(size=[1, 28, 28, 1]), training : False})    
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Mobile device

**Comment by ccurro** [comment]
Added

**Comment by wt-huang** [comment]
In the above code snippet you are performing Ghost Batch Normalization since virtual_batch_size is not None. This creates virtual sub-batches that are normalized separately with shared gamma, beta, and moving average. 

The main benefit of acquiring the statistics on small virtual batches instead of the real large batch is to reduce the generalization error. It is noted in the paper the full batch statistics need to be used for inference. This is the reason that you need to use multiples of virtual_batch_size for inference.

**Comment by ccurro** [comment]
Please look at the test phase of algorithm 1 in the [ghost norm paper](https://arxiv.org/pdf/1705.08741.pdf). The test phase requires the learned gamma and beta, whose shape are independent of the batch size. The moving statistics mu_run, and sigma_run are calculated "on the side" from the full batches during train time, and are otherwise not used during train time. Their size/shape is independent of the batch size. Only mu_run, sigma_run, gamma, and beta are used at test/inference phase, and are not dependent on the test/inference training examples. Test/inference should be able to happen for any batch_size.

Your suggestion implies that test inputs would not be treated independently, i.e. that it matters what they are batched with, which is not fundamentally sound.

**Comment by wt-huang** [comment]
Good that you dig deeper in the paper, guess I was not explicit in my explanation. 

> The test phase requires the learned gamma and beta, whose shape are independent of the batch size.

For normalization at inference time population mean and variance are used which were calculated during training using moving mean and variance. They are averaged over virtual batches. From Algorithm 1 in Hoffer's paper, I wouldn't say that alpha and gamma are independent of batch size. You can also find more details on Algorithm 1&2 in the paper from [Ioffe & Szegedy](https://arxiv.org/pdf/1502.03167.pdf).

> Test/inference should be able to happen for any batch_size.

This is true in general, but your last sess.run call is made on x_norm, i.e. tf.layers.batch_normalization of x during training where virtual_batch_size is used.

**Comment by ccurro** [comment]
@wt-huang I'm specifically saying their **shape** is independent of the batch size - of course their actual values depends on the batch size during training.

@ppwwyyxx +1

**Comment by wt-huang** [comment]
@ccurro  gamma and beta are parameters.

> > but your last sess.run call is made on x_norm, i.e. tf.layers.batch_normalization of x during training where virtual_batch_size is used.
> The last sess.run call has training=False.

Yes, but if you change virtual_batch in x_norm as below, sess.run would not incur error.
```
x_norm = tf.layers.batch_normalization(x, training=training, virtual_batch_size=1)
```

**Comment by ccurro** [comment]
I am aware that gamma and beta are parameters - nothing I have said indicates otherwise. 

Your statement is trivial and reductive. Obviously if I make `virtual_batch_size=1` then yes, at inference time, as the algorithm is currently implemented, I will not encounter a runtime error - that does not however make any statements about the correctness of the implementation. Further, `virtual_batch_size=1` is more than likely completely useless, since the variance for any virtual batch would be zero. 

Your statement does nothing to refute the utility, fundamental soundness, and clear expositional foundation (from the original paper) of training with a given batch size and evaluating with another, possibly singleton, batch size. 

Forward batch statistics are not computed during inference. There should be no requirement for an inference batch. All batch statistics are computed during training which are then used at inference time. 

If you will not change your position, please provide a sound argument or reassign this issue to another Googler/contributor.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by hckiang** [comment]
@ccurro Just run into the same problem and I completely agree with you that virtual_batch_size should only be respected in train mode. By now have you found any solutions to this yet?

**Comment by ccurro** [comment]
No solutions found

**Comment by Xiadalei** [comment]
It seems a long time has passed and the issue is still there (with TF2.0), any plan?

**Comment by reedwm** [comment]
@jhseu, can you handle this? I am unfamiliar with ghost batch norm.

**Comment by netanel-s** [comment]
@jhseu Is there any update regarding this issue?

**Comment by netanel-s** [comment]
@ccurro @hckiang @Xiadalei @jhseu 
I think I solved the bug, would appreciate you confirming.
I think if you replace in [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/normalization.py#L712), the line
`expanded_shape = [self.virtual_batch_size, -1] + original_shape[1:]`
with
`expanded_shape = control_flow_ops.cond(training,lambda:[self.virtual_batch_size, -1] + original_shape[1:], lambda:[1, -1] + original_shape[1:])`
It will behave as expected

(Need to add `from tensorflow.python.ops import control_flow_ops`).

**Comment by ccurro** [comment]
@netanel-s 

At a glance, your conditional looks like it would produce the correct behavior. Happy testing :smile: 

Thanks for checking it out :slightly_smiling_face:

**Comment by Pyrestone** [comment]
In general, can we make the behaviour of this option a bit more forgiving for uneven bactch sizes?
Maybe the following behaviour is more in accordance with the user's expectations:

In  **training** mode:
- If the batch size is a multiple of virtual_batch_size -> use virtual batch size.
- If the batch size is not a multiple -> use the whole batch size and raise a warning the first time this occurs.

In **testing** mode: No reshaping is needed as the running statistics are used.

Currently, tf raises an exception instead of a warning.
This change would allow users to use the model.fit() calls without worrying about the last batch being the right size or defaulting to a custom training loop.

**Comment by ghannum** [comment]
Is there any update on this? I attempted the fix from @netanel-s on TF2.3.1 but it hasn't fixed the issue for me. It's clear that virtual_batch_size should have no effect on tensor shapes for either training or inference..

**Comment by reedwm** [comment]
I'm marking as contributions welcome because no one currently has the bandwidth to implement this. If someone creates a PR for this fix, please CC me.

**Comment by w4nderlust** [comment]
Just wanted to say once more that @ccurro 's points are all entirely valid 3 years later.

**Comment by w4nderlust** [comment]
In the meantime, as a worksround, I implemented a GhostBatchNorm class in Ludwig that user BatchNorm internally. Posting it here as it may be useful: https://github.com/ludwig-ai/ludwig/blob/master/ludwig/modules/normalization_modules.py#L6-L38
